### PR TITLE
EZP-27492 Content draft edit

### PR DESCRIPTION
**Pull request by glye**
> Implement https://jira.ez.no/browse/EZP-27492
> Ready for review
> Based on https://github.com/ezsystems/repository-forms/pull/111 and https://github.com/ezsystems/repository-forms/pull/132

Preface:
- [x] speling fixes
- [x] fix merge conflict
- [x] fix CS issues
- [x] make tests pass

And then...
- [x] enable existing save draft / cancel actions, verify that they work
- [x] take out commits not strictly in scope (field type specific)
- [x] ~add edit button to content view~ Dealt with in other issues
- [x] ~add edit button to content version list, for specific version~ Dealt with in other issues

### CODE REVIEW

**Comment by glye** [comment]
@bdunogier Tests failing now because PHPunit 6.2 is installed for PHP7, rather than 5.7 as before. Known issue? Should be fixed in composer.json?

**Comment by bdunogier** [comment]
@glye if I am not mistaken, the package has been patched recently to support it. Can you have a look at the recent history, and see if there are relevant changes you should adapt to ?

**Comment by glye** [comment]
Huh, I already checked in master, but let's see..

**Comment by bdunogier** [comment]
Actually, you're right, Gunnstein. I also don't really see why it would install PhpUnit 6.x given the `composer.json` contents.

**Comment by glye** [comment]
You're right, there was such a recent change, but it didn't cover the `features`. I'll fix.

**Comment by glye** [comment]
The `features` should be phpunit 6.x compatible now. Similar change for other tests was done in the older https://github.com/ezsystems/repository-forms/commit/0290f1884924f1a500e73b1947251fa8c9a50308

**Comment by glye** [comment]
And we're green again 💚 
I'll reduce the number of commits.

**Comment by glye** [comment]
@bdunogier It's down to 5 commits, and ready for review if tests pass again. (Might also squash the 3rd commit...?)

File: lib/Data/Content/CreateContentDraftData.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\RepositoryForms\Data\Content;
+
+use eZ\Publish\API\Repository\Values\ValueObject;
+
+class CreateContentDraftData extends ValueObject
+{
+    public $contentId;
```
Comment (bdunogier) [inline lib/Data/Content/CreateContentDraftData.php]: In retrospect, this isn't very value objectish. In our ValueObjects, properties are meant to be protected, and set using the constructor (read only VO).

Could you update the properties for that, and adapt the action ?

**Comment by glye** [comment]
Squashed all except the PHPunit 6.x fix.

File: lib/Data/Content/CreateContentDraftData.php (hunk 1 above)
Comment (glye) [inline lib/Data/Content/CreateContentDraftData.php]: @bdunogier Symfony `PropertyPathMapper::mapFormsToData()` doesn't like that, it needs to write to the data.
http://api.symfony.com/3.1/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.html#method_mapFormsToData
Could workaround that with setter methods, but does it make sense?

**Comment by glye** [comment]
Squashed all except the PHPunit 6.x fix, and the field value validator change.

File: lib/Data/Content/CreateContentDraftData.php (hunk 1 above)
Comment (glye) [inline lib/Data/Content/CreateContentDraftData.php]: @bdunogier We could maybe also "wrap" a readonly VO in a writeable VO for the purpose of the mapper, but again, does this make sense to do?

**Comment by glye** [comment]
@bdunogier I pushed this fixup just to illustrate the problem. This gathers the parameters, uses them in the `CreateContentDraftData` constructor, and uses the value object with the form, the way I interpreted your suggestion. It fails with a `PropertyReadOnlyException` due to `PropertyPathMapper::mapFormsToData()` wanting to edit properties.

Perhaps it can be solved by reimplementing the mapper, or something like that. I haven't had time to look much into it yet.

File: lib/Data/Content/CreateContentDraftData.php (hunk 1 above)
Comment (bdunogier) [inline lib/Data/Content/CreateContentDraftData.php]: Please disregard this. It is valid for values that are set in the repository. This is input data, meant to be manipulated by the client (the mapper here).

**Comment by glye** [comment]
Reverted readonly valueobject commit after discussion with @bdunogier.

File: doc/specifications/drafts/content_edit.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
+# Symfony forms content edit
```
Comment (bdunogier) [inline doc/specifications/drafts/content_edit.md]: I wouldn't merge this document as is.

Could you try to update it / write a shorter one that covers what is *actually* in *this* PR ?

**Comment by glye** [comment]
@bdunogier Update: Trimmed down doc, and squashed. New doc in [doc/specifications/content_edit.md](https://github.com/ezsystems/repository-forms/pull/133/files#diff-2f0483f64cc589f00c1366085d8f17c6)

File: doc/specifications/content_edit.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+editing with drafts.
+
+Editing content implies creating a new version of an existing content, rendering a form with the
+previous version's fields values, and allowing the user to save the draft, cancel or publish the
+new version.
+
+
+### Creating a draft from an existing version
+
+> A route that displays a form allowing to create a draft from an existing content item.
```
Comment (bdunogier) [inline doc/specifications/content_edit.md]: Missing word ?

Comment (glye) [inline doc/specifications/content_edit.md]: You mean "...allowing *you* to..."? It's common and works fine either way I think. I wouldn't change it unless a native speaker says it's borked :)

**Pull request merged**